src/workflow_dispatch.ts

```diff
@@ -3,56 +3,137 @@ import * as core from '@actions/core'
+  const caller: Caller = { octokit }
+
+  // Normalize owner, repo and ref (if any)
   const ref = opts.ref || github.context.ref
   const [owner, repo] = opts.repo
-    ? opts.repo.split('/')
+    ? (opts.repo as string).split('/')
     : [github.context.repo.owner, github.context.repo.repo]
 
-  const octokit = new Octokit({ auth: opts.token })
+  Object.assign(caller, opts, { owner, repo, ref })
+  return caller
+}
 
-  type WorkflowsResponseList =
-    Endpoints['GET /repos/{owner}/{repo}/actions/workflows']['response']['data']['workflows']
-  const workflows: WorkflowsResponseList = await octokit.paginate(
-    octokit.rest.actions.listRepoWorkflows.endpoint.merge({
-      owner,
-      repo,
-      ref,
-      inputs
+// Find a workflow in the repo
+async function getWorkflow(caller: Caller): Promise<WorkflowsList[0]> {
+  let found
+  const workflows: WorkflowsList = await caller.octokit.paginate(
+    caller.octokit.rest.actions.listRepoWorkflows.endpoint.merge({
+      owner: caller.owner,
+      repo: caller.repo
     })
   )
 
-  let found
   for (const w of workflows) {
-    if (w.name === opts.workflow || w.id.toString() === opts.workflow) {
+    if (w.name === caller.workflow || w.id.toString() === caller.workflow) {
       found = w
+      core.debug(`Dispatched workflow ${caller.workflow} (${found.id})`)
       break
     }
   }
 
   if (!found) {
-    throw new Error(
-      `Unable to find workflow '${opts.workflow}' in ${owner}/${repo} 😥`
-    )
+    throw new Error(`Unable to find workflow: '${caller.workflow}' 😥`)
   }
-  core.debug(`Dispatch workflow ${opts.workflow} (${found.id})`)
-  const dispatchResp = await octokit.request(
-    `POST /repos/${owner}/${repo}/actions/workflows/${found.id}/dispatches`,
+
+  return found
+}
+
+async function getLatestRuns(
+  workflow: WorkflowsList[0],
+  caller: Caller,
+  latest_number = 10
+): Promise<WorkflowsRunsList> {
+  const options = caller.octokit.rest.actions.listWorkflowRuns.endpoint.merge({
+    owner: caller.owner,
+    repo: caller.repo,
+    workflow_id: workflow.id,
+    per_page: latest_number
+  })
+
+  const runs = await caller.octokit.paginate(options, (response, done) => {
+    done()
+    return response.data as WorkflowsRunsList
+  })
+
+  return runs
+}
+
+// Wait for queued
+async function waitForQueued(
+  workflow: WorkflowsList[0],
+  headRun: WorkflowsRunsList[0],
+  caller: Caller
+): Promise<WorkflowsRunsList[0] | undefined> {
+  const expected = ['in_progress', 'queued', 'requested', 'waiting']
+  let run_number = 0
+  if (headRun) {
+    run_number = headRun.run_number
+  }
+
+  // We actually admit to catch one of the expected statuses in the following ten seconds
+  // In other words we fail to detect the workflow run if something might go wrong
+  for (let i = 0; i < 10; i++) {
+    // sleep for a second
+    await new Promise(resolve => setTimeout(resolve, 1000))
+    const runs = await getLatestRuns(workflow, caller, 10)
```

Comment:
so we expect the workflow to kick in within 10 seconds?

Reply:
Yup, that's right.

Reply:
sometimes takes longer. I suggest a bit more patience.